std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,247 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range.
+This function is $(BIGOH r.length).
+
+Params:
+    r = an input range
+
+Returns:
+    the mean of r
+*/
+auto mean(R)(R r)
+if (isInputRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    static if (hasLength!R)
+    {
+        if (r.length == 0)
+        {
+            return double.nan;
+        }
+        else if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r) / r.length;
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+            (tuple(size_t(0), 0.0), r);
+
+
+        if (pair[0] == 0)
+        {
+            return double.nan;
+        }
+
+        return pair[1] / pair[0];
+    }
+}
+
+///
+@safe pure nothrow unittest
+{
+    import std.math : approxEqual;
+
+    assert(mean([1, 2, 3]) == 2);
+    assert(approxEqual(mean([1.5, 2.5, 12.5]), 5.5));
+}
+
+@safe pure nothrow unittest
+{
+    import std.internal.test.dummyrange : ReferenceInputRange;
+    import std.math : approxEqual;
+
+    auto r1 = new ReferenceInputRange!int([1, 2, 3]);
+    assert(r1.mean == 2);
+
+    auto r2 = new ReferenceInputRange!double([1.5, 2.5, 12.5]);
+    assert(approxEqual(r2.mean, 5.5));
+}
+
+/**
+Function to compute the quantile of a $(D SortedRange) given $(D q), which
+is a number between 0 and 1. If the range is empty, this function will fail
+an assert in debug mode and return NaN in release mode. Is $(BIGOH r.length * q)
+for forward ranges and $(BIGOH 1) for random access ranges.
+
+Params:
+    r = a forward range with length or a random access range
+    q = where to calculate the quantile
+
+Returns:
+    the quantile of the range given $(D q)
+*/
+double quantile(alias less = "a < b", R)(SortedRange!(R, less) r, double q)
```

Comment:
There is still the concern over what the `less` parameter means, in terms of the returned value. What is the result of `r.quantile!`a > b`(0.25)`, for example? What does it mean, numerically? What *should* it mean?

What I'm trying to get at, is, does it even make any sense for `less` to be anything other than `a < b`? If `less` is any other ordering than `a < b`, will this function actually compute the quantile, or will it compute something completely different? If it doesn't make sense for `less` to be anything other than `a < b`, why is it a parameter that can be specified by the user?